# BAFU dataset extraction

Author: Thiago Nascimento (thiago.nascimento@eawag.ch)

This notebook is used to retrieve and concatenate the NADUF dataset. They present indeed different time-resolutions so not necessarly daily and hourly, but in different campaings. 

The output is one file per catchemnt (similar to the CAMELS_CH), with 44 columns:

* mean_discharge(m3/s)
* total_discharge(Mio m3)
* temperature_BAFU(°C)	
* pH_BAFU()	
* conductivity_25C_BAFU(µS/cm)	
* oxygen(mg/l)	
* oxygen_saturation(%)	
* pH_lab()	
* conductivity_20C_lab(µS/cm)
* total_hardness(mmol/l)
* alkalinity(mmol/l)	
* calcium(mg/l)	
* magnesium(mg/l)	
* nitrate(mg N/l)	
* total_nitrogen(mg N/l)	
* DRP(mg P/l)	
* total_phosphorus(mg P/l)	
* total_phosphorus_filtered(mg P/l)	
* chloride(mg/l)	
* fluoride(mg/l)	
* bromide(mg/l)	
* silicate(mg H4SiO4/l)	
* sulphate(mg SO4/l)	
* sodium(mg/l)	
* potassium(mg/l)	
* iron(mg/l)	
* TOC(mg C/l)	
* DOC(mg C/l)	
* suspended_material(mg/l)	
* chromium(µg/l)	
* zinc(µg/l)	
* copper(µg/l)	
* cadmium(µg/l)	
* lead(µg/l)	
* nickel(µg/l)	
* mercury(µg/l)	
* barium(µg/l)	
* strontium(µg/l)	
* arsenic(µg/l)	
* manganese(µg/l)	
* NP(µg/l)	
* NP1EO(µg/l)	
* NP2EO(µg/l)	
* NP3EO(µg/l)

## Requirements
**Python:**

* Python>=3.6
* Jupyter
* geopandas=0.10.2
* numpy
* os
* pandas=2.1.3
* scipy=1.9.0
* tqdm

Check the Github repository for an environment.yml (for conda environments) or requirements.txt (pip) file.

**Files:**

* 


**Directory:**

* Clone the GitHub directory locally
* Place any third-data variables in their respective directory.
* ONLY update the "PATH" variable in the section "Configurations", with their relative path to the EStreams directory. 


## References
* 
## Observations
* 

# Import modules

In [1]:
import pandas as pd
import geopandas as gpd
import numpy as np
import tqdm as tqdm
import os
import glob
import warnings
import re

# Configurations

In [2]:
# Only editable variables:
# Relative path to your local directory
PATH = ".."
# Suppress all warnings
warnings.filterwarnings("ignore")

* #### The users should NOT change anything in the code below here. 

In [3]:
# Non-editable variables:
PATH_OUTPUT = "results/interval_samples/"

# Set the directory:
os.chdir(PATH)

In [4]:
os.getcwd()

'c:\\Users\\nascimth\\Documents\\Thiago\\Eawag\\Python\\Scripts\\CAMELS_CH_chem'

# Import data
* NADUF

In [13]:
# Full dataset of interval (time-series)
path = r'data\NADUF\\'

filenames = glob.glob(path + "*.xlsx")
len(filenames)

1

In [12]:
# Network NADUF
network_naduf = pd.read_excel(r"data/CAMELS_CH_chem_stations_short_v2.xlsx", sheet_name='naduf')
network_naduf

naduf_id             naduf_station naduf_water_body  longitude_LV03  \
0       1837             Porte du Scex            Rhône          557660   
1       1833                     Brugg             Aare          657000   
2       1835                 Mellingen            Reuss          662830   
3       1823               Andelfingen             Thur          693510   
4       1842                   Martina              Inn          830640   
5       1840                  Riazzino           Ticino          713670   
6       1832                   Hagneck             Aare          580680   
7       1827  Rheinfelden, Messstation            Rhein          627190   
8       1828     Münchenstein, Hofmatt             Birs          613570   
9       4409                 Appenzell           Sitter          749040   
10      1826                Laufenburg            Rhein          646000   
11      1830             Bern, Schönau             Aare          600710   
12      1825                  Rekingen            Rhein          667060   
13      1838         Chancy, Aux Ripes            Rhône          486600   
14      3717     Thörishaus, Sensematt            Sense          593350   
15      1836                Gebenstorf           Limmat          660530   
16      4879                Frauenfeld             Murg          709540   
17      1824              Rheinsfelden            Glatt          678040   
18      2064                   S-chanf              Inn          795800   
19      1831                  Gümmenen            Saane          585100   
20      1821    Diepoldsau, Rietbrücke            Rhein          766280   
21      1246                Neuenkirch  Sellenbodenbach          658530   
22      2078      Weil, Palmrainbrücke            Rhein          611740   
23      6169                     Emmen      Kleine Emme          663700   

    lattitude_LV03  area_camels_CH [km2]  hydro_naduf_distance [km]  \
0           133280                5239.4                        0.0   
1           259360               11681.3                        0.0   
2           252580                3385.8                        0.0   
3           272500                1701.6                        0.0   
4           197190                1937.5                        0.0   
5           113500                1613.3                       10.2   
6           211650                5111.9                        0.0   
7           267840               34479.4                        0.0   
8           263080                 887.3                        0.0   
9           244220                  74.4                        0.0   
10          267410               34010.3                        0.0   
11          198000                2941.2                        0.0   
12          269230               14719.1                        0.0   
13          112340               10312.7                        0.0   
14          193020                 351.2                        0.0   
15          260770                2412.0                        7.1   
16          269660                 213.3                        0.0   
17          269720                 417.4                        0.0   
18          165910                 616.0                        3.0   
19          199240                1881.3                        5.3   
20          250360                6297.4                        0.0   
21          218290                  10.4                        0.0   
22          272310               36404.4                        7.2   
23          213630                 478.3                        0.0   

    waterqual_id  
0           2009  
1           2016  
2           2018  
3           2044  
4           2067  
5           2068  
6           2085  
7           2091  
8           2106  
9           2112  
10          2130  
11          2135  
12          2143  
13          2174  
14          2179  
15          2622  
16          2386  
17        

In [6]:
dataset_naduf = pd.read_excel(r"data/NADUF/naduf_data_1981-2020_v4.xlsx")
dataset_naduf

naduf_id status_number                                          remark  \
0          1181             1                                             NaN   
1          1181             1                                             NaN   
2          1181             1                                             NaN   
3          1181             1                                             NaN   
4          1181             1                                             NaN   
...         ...           ...                                             ...   
14464      6169             1                                             NaN   
14465      6169             1                                             NaN   
14466      6169             1                                             NaN   
14467      6169             1                                             NaN   
14468      6169             9  "1. Woche: Ankunft im Labor erst am 16.12.20."   

       year            date_end    duration  mean_discharge  total_discharge  \
0      1982 1982-11-15 06:00:00  336.000000        5.313851         6.427634   
1      1982 1982-11-29 06:00:00  336.000000        9.046227        10.942316   
2      1982 1982-12-13 05:30:00  335.000000       10.864181        13.102202   
3      1982 1982-12-27 05:30:00  336.000000       27.653205        33.449317   
4      1983 1983-01-10 05:55:00  337.000000       12.789252        15.515920   
...     ...                 ...         ...             ...              ...   
14464  2020 2020-10-26 10:35:00  336.833333       10.570123        12.817331   
14465  2020 2020-11-09 10:15:00  335.666667       15.549123        18.789560   
14466  2020 2020-11-23 10:05:00  335.833333        6.439759         7.785669   
14467  2020 2020-12-07 10:05:00  336.000000        4.323554         5.229770   
14468  2020 2020-12-21 15:15:00  341.166667        8.327426        10.227744   

       temperature_BAFU  pH_BAFU  ...       zinc    copper   cadmium  \
0              7.271006      NaN  ...  10.406677  3.019255  0.073556   
1              4.826679      NaN  ...        NaN       NaN       NaN   
2              4.872490      NaN  ...        NaN       NaN       NaN   
3              3.792989      NaN  ...   26.11415  2.798573  0.025071   
4              3.064523      NaN  ...  28.858242  2.764272  0.026786   
...                 ...      ...  ...        ...       ...       ...   
14464          8.946462      NaN  ...        NaN       NaN       NaN   
14465          8.916996      NaN  ...        NaN       NaN       NaN   
14466          6.657077      NaN  ...        NaN       NaN       NaN   
14467          3.561146      NaN  ...        NaN       NaN       NaN   
14468          3.696536      NaN  ...        NaN       NaN       NaN   

           lead  nickel  mercury  barium  strontium arsenic manganese  
0      1.790373     NaN      NaN     NaN        NaN     NaN       NaN  
1           NaN     NaN      NaN     NaN        NaN     NaN       NaN  
2           NaN     NaN      NaN     NaN        NaN     NaN       NaN  
3      1.349287     NaN      NaN     NaN        NaN     NaN       NaN  
4      1.332136     NaN      NaN     NaN        NaN     NaN       NaN  
...         ...     ...      ...     ...        ...     ...       ...  
14464       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14465       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14466       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14467       NaN     NaN      NaN     NaN        NaN     NaN       NaN  
14468       NaN     NaN      NaN     NaN        NaN     NaN       NaN  

[14469 rows x 46 columns]

In [8]:
len(dataset_naduf.naduf_id.unique())

29

In [9]:
dataset_naduf.naduf_id.unique()

array([1181, 1246, 1821, 1822, 1823, 1824, 1825, 1826, 1828, 1829, 1830,
       1831, 1832, 1833, 1835, 1836, 1837, 1838, 1840, 1842, 2044, 2045,
       2046, 2064, 2078, 3717, 4409, 4879, 6169], dtype=int64)

In [38]:
column_rename_dict = {
    'naduf_id': 'naduf_id', 
    'status_number': 'status_number', 
    'remark':'remark' , 
    'year':'year', 
    'date_end':'date', 
    'duration':'duration',
    'mean_discharge': 'mean_discharge(m3/s)',
    'total_discharge': 'total_discharge(Miom3)',
    'temperature_BAFU': 'temperature_BAFU(°C)',
    'pH_BAFU': 'pH_BAFU(-)',
    'conductivity_25C_BAFU': 'conductivity_25C_BAFU(µS/cm)',
    'oxygen': 'oxygen(mg/l)',
    'oxygensaturation': 'oxygen_saturation(%)',
    'pH_lab': 'pH_lab(-)',
    'conductivity_20C_lab': 'conductivity_20C_lab(µS/cm)',
    'total_hardness': 'total_hardness(mmol/l)',
    'alkalinity': 'alkalinity(mmol/l)',
    'calcium': 'calcium(mg/l)',
    'magnesium': 'magnesium(mg/l)',
    'nitrate': 'nitrate(mgN/l)',
    'total_nitrogen': 'total_nitrogen(mgN/l)',
    'DRP': 'DRP(mgP/l)',
    'total_phosphorus': 'total_phosphorus(mgP/l)',
    'total_phosphorus_filtered': 'total_phosphorus_filtered(mgP/l)',
    'chloride': 'chloride(mg/l)',
    'fluoride': 'fluoride(mg/l)',
    'bromide': 'bromide(mg/l)',
    'silicate': 'silicate(mgH4SiO4/l)',
    'sulphate': 'sulphate(mgSO4/l)',
    'sodium': 'sodium(mg/l)',
    'potassium': 'potassium(mg/l)',
    'iron': 'iron(mg/l)',
    'TOC': 'TOC(mgC/l)',
    'DOC': 'DOC(mgC/l)',
    'suspended_material': 'suspended_material(mg/l)',
    'chromium': 'chromium(µg/l)',
    'zinc': 'zinc(µg/l)',
    'copper': 'copper(µg/l)',
    'cadmium': 'cadmium(µg/l)',
    'lead': 'lead(µg/l)',
    'nickel': 'nickel(µg/l)',
    'mercury': 'mercury(µg/l)',
    'barium': 'barium(µg/l)',
    'strontium': 'strontium(µg/l)',
    'arsenic': 'arsenic(µg/l)',
    'manganese': 'manganese(µg/l)'
}

In [39]:
# Rename columns based on the dictionary
dataset_naduf.rename(columns=column_rename_dict, inplace=True)

In [40]:
dataset_naduf.columns

Index(['naduf_id', 'status_number', 'remark', 'year', 'date', 'duration',
       'mean_discharge(m3/s)', 'total_discharge(Miom3)',
       'temperature_BAFU(°C)', 'pH_BAFU(-)', 'conductivity_25C_BAFU(µS/cm)',
       'oxygen(mg/l)', 'oxygen_saturation(%)', 'pH_lab(-)',
       'conductivity_20C_lab(µS/cm)', 'total_hardness(mmol/l)',
       'alkalinity(mmol/l)', 'calcium(mg/l)', 'magnesium(mg/l)',
       'nitrate(mgN/l)', 'total_nitrogen(mgN/l)', 'DRP(mgP/l)',
       'total_phosphorus(mgP/l)', 'total_phosphorus_filtered(mgP/l)',
       'chloride(mg/l)', 'fluoride(mg/l)', 'bromide(mg/l)',
       'silicate(mgH4SiO4/l)', 'sulphate(mgSO4/l)', 'sodium(mg/l)',
       'potassium(mg/l)', 'iron(mg/l)', 'TOC(mgC/l)', 'DOC(mgC/l)',
       'suspended_material(mg/l)', 'chromium(µg/l)', 'zinc(µg/l)',
       'copper(µg/l)', 'cadmium(µg/l)', 'lead(µg/l)', 'nickel(µg/l)',
       'mercury(µg/l)', 'barium(µg/l)', 'strontium(µg/l)', 'arsenic(µg/l)',
       'manganese(µg/l)'],
      dtype='object')

In [41]:
dataset_naduf = dataset_naduf[['naduf_id', 'date',
       'mean_discharge(m3/s)',
       'temperature_BAFU(°C)', 'pH_BAFU(-)', 'conductivity_25C_BAFU(µS/cm)',
       'oxygen(mg/l)', 'oxygen_saturation(%)', 'pH_lab(-)',
       'conductivity_20C_lab(µS/cm)', 'total_hardness(mmol/l)',
       'alkalinity(mmol/l)', 'calcium(mg/l)', 'magnesium(mg/l)',
       'nitrate(mgN/l)', 'total_nitrogen(mgN/l)', 'DRP(mgP/l)',
       'total_phosphorus(mgP/l)', 'total_phosphorus_filtered(mgP/l)',
       'chloride(mg/l)', 'fluoride(mg/l)', 'bromide(mg/l)',
       'silicate(mgH4SiO4/l)', 'sulphate(mgSO4/l)', 'sodium(mg/l)',
       'potassium(mg/l)', 'iron(mg/l)', 'TOC(mgC/l)', 'DOC(mgC/l)',
       'suspended_material(mg/l)', 'chromium(µg/l)', 'zinc(µg/l)',
       'copper(µg/l)', 'cadmium(µg/l)', 'lead(µg/l)', 'nickel(µg/l)',
       'mercury(µg/l)', 'barium(µg/l)', 'strontium(µg/l)', 'arsenic(µg/l)',
       'manganese(µg/l)']]
dataset_naduf

naduf_id                date  mean_discharge(m3/s)  \
0          1181 1982-11-15 06:00:00              5.313851   
1          1181 1982-11-29 06:00:00              9.046227   
2          1181 1982-12-13 05:30:00             10.864181   
3          1181 1982-12-27 05:30:00             27.653205   
4          1181 1983-01-10 05:55:00             12.789252   
...         ...                 ...                   ...   
14464      6169 2020-10-26 10:35:00             10.570123   
14465      6169 2020-11-09 10:15:00             15.549123   
14466      6169 2020-11-23 10:05:00              6.439759   
14467      6169 2020-12-07 10:05:00              4.323554   
14468      6169 2020-12-21 15:15:00              8.327426   

       temperature_BAFU(°C)  pH_BAFU(-)  conductivity_25C_BAFU(µS/cm)  \
0                  7.271006         NaN                           NaN   
1                  4.826679         NaN                           NaN   
2                  4.872490         NaN                           NaN   
3                  3.792989         NaN                           NaN   
4                  3.064523         NaN                           NaN   
...                     ...         ...                           ...   
14464              8.946462         NaN                           NaN   
14465              8.916996         NaN                           NaN   
14466              6.657077         NaN                           NaN   
14467              3.561146         NaN                           NaN   
14468              3.696536         NaN                           NaN   

       oxygen(mg/l)  oxygen_saturation(%)  pH_lab(-)  \
0               NaN                   NaN   8.480000   
1               NaN                   NaN   8.430766   
2               NaN                   NaN   8.356344   
3               NaN                   NaN   8.474643   
4               NaN                   NaN   8.281437   
...             ...                   ...        ...   
14464           NaN                   NaN   8.410000   
14465           NaN                   NaN   8.390000   
14466           NaN                   NaN   8.380000   
14467           NaN                   NaN   8.400031   
14468           NaN                   NaN   8.414994   

       conductivity_20C_lab(µS/cm)  ...  zinc(µg/l)  copper(µg/l)  \
0                       367.066770  ...   10.406677      3.019255   
1                       356.188458  ...         NaN           NaN   
2                       348.381529  ...         NaN           NaN   
3                       338.992271  ...    26.11415      2.798573   
4                       274.674792  ...   28.858242      2.764272   
...                            ...  ...         ...           ...   
14464                   317.222586  ...         NaN           NaN   
14465                   313.269589  ...         NaN           NaN   
14466                   342.636797  ...         NaN           NaN   
14467                   383.225945  ...         NaN           NaN   
14468                   368.479080  ...         NaN           NaN   

       cadmium(µg/l) lead(µg/l) nickel(µg/l) mercury(µg/l) barium(µg/l)  \
0           0.073556   1.790373          NaN           NaN          NaN   
1                NaN        NaN          NaN           NaN          NaN   
2                NaN        NaN          NaN           NaN          NaN   
3           0.025071   1.349287          NaN           NaN          NaN   
4           0.026786   1.332136          NaN           NaN          NaN   
...              ...        ...          ...           ...          ...   
14464            NaN        NaN          NaN           NaN          NaN   
14465            NaN        NaN          NaN           NaN          NaN   
14466            NaN        NaN          NaN           NaN          NaN   
14467            NaN        NaN          NaN           NaN          NaN   
14468            NaN        NaN          NaN           NaN          Na

In [48]:
# Convert to datetime:
dataset_naduf["date"] = pd.to_datetime(dataset_naduf["date"], format='%Y-%m-%d')

# Assign the new code:
dataset_naduf["gauge_id"] = "NADU" + dataset_naduf["naduf_id"].astype(str)
dataset_naduf.head()

naduf_id                date  mean_discharge(m3/s)  temperature_BAFU(°C)  \
0      1181 1982-11-15 06:00:00              5.313851              7.271006   
1      1181 1982-11-29 06:00:00              9.046227              4.826679   
2      1181 1982-12-13 05:30:00             10.864181              4.872490   
3      1181 1982-12-27 05:30:00             27.653205              3.792989   
4      1181 1983-01-10 05:55:00             12.789252              3.064523   

   pH_BAFU(-)  conductivity_25C_BAFU(µS/cm)  oxygen(mg/l)  \
0         NaN                           NaN           NaN   
1         NaN                           NaN           NaN   
2         NaN                           NaN           NaN   
3         NaN                           NaN           NaN   
4         NaN                           NaN           NaN   

   oxygen_saturation(%)  pH_lab(-)  conductivity_20C_lab(µS/cm)  ...  \
0                   NaN   8.480000                   367.066770  ...   
1                   NaN   8.430766                   356.188458  ...   
2                   NaN   8.356344                   348.381529  ...   
3                   NaN   8.474643                   338.992271  ...   
4                   NaN   8.281437                   274.674792  ...   

   copper(µg/l)  cadmium(µg/l)  lead(µg/l) nickel(µg/l) mercury(µg/l)  \
0      3.019255       0.073556    1.790373          NaN           NaN   
1           NaN            NaN         NaN          NaN           NaN   
2           NaN            NaN         NaN          NaN           NaN   
3      2.798573       0.025071    1.349287          NaN           NaN   
4      2.764272       0.026786    1.332136          NaN           NaN   

  barium(µg/l) strontium(µg/l) arsenic(µg/l) manganese(µg/l)  gauge_id  
0          NaN             NaN           NaN             NaN  NADU1181  
1          NaN             NaN           NaN             NaN  NADU1181  
2          NaN             NaN           NaN             NaN  NADU1181  
3          NaN             NaN           NaN             NaN  NADU1181  
4          NaN             NaN           NaN             NaN  NADU1181  

[5 rows x 42 columns]

In [49]:
# Retrieve the dataset names
naduf_codes = dataset_naduf["gauge_id"].unique().tolist()
naduf_codes

['NADU1181',
 'NADU1246',
 'NADU1821',
 'NADU1822',
 'NADU1823',
 'NADU1824',
 'NADU1825',
 'NADU1826',
 'NADU1828',
 'NADU1829',
 'NADU1830',
 'NADU1831',
 'NADU1832',
 'NADU1833',
 'NADU1835',
 'NADU1836',
 'NADU1837',
 'NADU1838',
 'NADU1840',
 'NADU1842',
 'NADU2044',
 'NADU2045',
 'NADU2046',
 'NADU2064',
 'NADU2078',
 'NADU3717',
 'NADU4409',
 'NADU4879',
 'NADU6169']

In [60]:
for code in tqdm.tqdm(naduf_codes):
    
    dataset_naduf[dataset_naduf["gauge_id"] == code]
    dataset = dataset_naduf[dataset_naduf["gauge_id"] == code]
    dataset.set_index("date", inplace = True)
    dataset.drop(["naduf_id", "gauge_id"], axis=1, inplace = True)
    
    dataset.index.name = "date"
    
    dataset = dataset.round(2)

    dataset.to_csv(PATH_OUTPUT + "/CAMELS_CH_chem_intervals_"+code+".csv", encoding='latin')
    

100%|██████████████████████████████████████████████████████████████████████████████████| 29/29 [00:00<00:00, 34.52it/s]


Here we have only 2622 (2243). 

### Aggregation

* We have 90 stations in total.
* The 88 from the temperature list + 2290 + 2176.
* We first aggregate the data avaialble from BAFU, then the 2 extra stations.

#### - BAFU

In [49]:
columns_names = ["temperature(°C)", "pH(-)", "electric_cond(µS/cm)", "oxygen_conc(mg/l)"]
hourly_timeseries = pd.DataFrame(index = pd.date_range('01-01-1970','12-31-2022', freq='H'), columns=columns_names)

for filename in tqdm.tqdm(temperature_filenames):

    network_filename = pd.read_csv(filename, skiprows=8, delimiter = ";", encoding='latin-1', usecols=["Zeitstempel", 
                                                                                                      "Wert"])
    network_filename["date"] = pd.to_datetime(network_filename["Zeitstempel"], format='%Y-%m-%d')
    network_filename = network_filename[["date", "Wert"]]
    network_filename.set_index("date", inplace = True)
    
    namestation = os.path.basename(filename)
    namestation = namestation.split("_", 5)[0]
    hourly_timeseries["temperature(°C)"] = network_filename["Wert"]
    
    
    ###################################################
    ##  pH
    
    filenames = ph_filenames.copy()
    # List to store filenames containing the number
    matching_filenames = []

    # Iterate over filenames and check if the number is present in the first 4 characters of the file name
    for filename in filenames:
        # Extract the file name (without the path)
        file_name_only = filename.split("\\")[-1]
        # Extract the first 4 characters of the file name
        first_four_characters = file_name_only[:4]
        # Check if the number is present in the first 4 characters
        if namestation in first_four_characters:
            matching_filenames.append(filename)
    try:        
        network_filename = pd.read_csv(matching_filenames[0], skiprows=8, delimiter = ";", encoding='latin-1', usecols=["Zeitstempel", 
                                                                                                          "Wert"])
        network_filename["date"] = pd.to_datetime(network_filename["Zeitstempel"], format='%Y-%m-%d')
        network_filename = network_filename[["date", "Wert"]]
        network_filename.set_index("date", inplace = True)
        network_filename['Wert'] = pd.to_numeric(network_filename['Wert'], errors='coerce')

        hourly_timeseries["pH(-)"] = network_filename["Wert"]      
    except: 
        1+1
        
    ###################################################
    ##  electric_cond(µS/cm)
    
    filenames = electric_filenames.copy()
    # List to store filenames containing the number
    matching_filenames = []

    # Iterate over filenames and check if the number is present in the first 4 characters of the file name
    for filename in filenames:
        # Extract the file name (without the path)
        file_name_only = filename.split("\\")[-1]
        # Extract the first 4 characters of the file name
        first_four_characters = file_name_only[:4]
        # Check if the number is present in the first 4 characters
        if namestation in first_four_characters:
            matching_filenames.append(filename)
    try:       
        network_filename = pd.read_csv(matching_filenames[0], skiprows=8, delimiter = ";", encoding='latin-1', usecols=["Zeitstempel", 
                                                                                                          "Wert"])
        network_filename["date"] = pd.to_datetime(network_filename["Zeitstempel"], format='%Y-%m-%d')
        network_filename = network_filename[["date", "Wert"]]
        network_filename.set_index("date", inplace = True)
        network_filename['Wert'] = pd.to_numeric(network_filename['Wert'], errors='coerce')

        hourly_timeseries["electric_cond(µS/cm)"] = network_filename["Wert"]
    except:
        1+1
    ###################################################
    ##  oxygen_conc(mg/l)
    
    filenames = oxygen_filenames.copy()
    # List to store filenames containing the number
    matching_filenames = []

    # Iterate over filenames and check if the number is present in the first 4 characters of the file name
    for filename in filenames:
        # Extract the file name (without the path)
        file_name_only = filename.split("\\")[-1]
        # Extract the first 4 characters of the file name
        first_four_characters = file_name_only[:4]
        # Check if the number is present in the first 4 characters
        if namestation in first_four_characters:
            matching_filenames.append(filename)
    try:      
        network_filename = pd.read_csv(matching_filenames[0], skiprows=8, delimiter = ";", encoding='latin-1', usecols=["Zeitstempel", 
                                                                                                          "Wert"])
        network_filename["date"] = pd.to_datetime(network_filename["Zeitstempel"], format='%Y-%m-%d')
        network_filename = network_filename[["date", "Wert"]]
        network_filename.set_index("date", inplace = True)
        network_filename['Wert'] = pd.to_numeric(network_filename['Wert'], errors='coerce')

        hourly_timeseries["oxygen_conc(mg/l)"] = network_filename["Wert"]  
    except: 
        1+1
    
    hourly_timeseries.index.name = "date"
    
    hourly_timeseries = hourly_timeseries.round(2)

    hourly_timeseries.to_csv(PATH_OUTPUT + "/hourly/CAMELS_CH_chem_hourly_"+namestation+".csv", encoding='latin')

100%|██████████████████████████████████████████████████████████████████████████████████| 88/88 [03:50<00:00,  2.62s/it]


#### - 2 extra stations
Station 2176

In [74]:
namestation = "2176"

columns_names = ["temperature(°C)", "pH(-)", "electric_cond(µS/cm)", "oxygen_conc(mg/l)"]
hourly_timeseries = pd.DataFrame(index = pd.date_range('01-01-1970','12-31-2022', freq='H'), columns=columns_names)

path_2176 = 'data\Kanton ZH\S2176\S2176_Wassertemperatur _Stundenmittel.txt'
network_2176 = pd.read_csv(path_2176, skiprows=2, delimiter = "\t", encoding='latin-1')
network_2176["date"] = network_2176["Datum"] + " "  + network_2176["Zeit"]
network_2176["date"] = pd.to_datetime(network_2176["date"], format='%d.%m.%Y %H:%M:%S')
network_2176.set_index("date", inplace = True)
network_2176 = network_2176[["Messwert"]]
hourly_timeseries["temperature(°C)"] = network_2176["Messwert"]

hourly_timeseries.index.name = "date"
    
hourly_timeseries = hourly_timeseries.round(2)

hourly_timeseries.to_csv(PATH_OUTPUT + "/hourly/CAMELS_CH_chem_hourly_"+namestation+".csv", encoding='latin')

hourly_timeseries

temperature(°C) pH(-) electric_cond(µS/cm)  \
date                                                              
1970-01-01 00:00:00              NaN   NaN                  NaN   
1970-01-01 01:00:00              NaN   NaN                  NaN   
1970-01-01 02:00:00              NaN   NaN                  NaN   
1970-01-01 03:00:00              NaN   NaN                  NaN   
1970-01-01 04:00:00              NaN   NaN                  NaN   
...                              ...   ...                  ...   
2022-12-30 20:00:00             6.49   NaN                  NaN   
2022-12-30 21:00:00             6.51   NaN                  NaN   
2022-12-30 22:00:00             6.55   NaN                  NaN   
2022-12-30 23:00:00             6.59   NaN                  NaN   
2022-12-31 00:00:00             6.63   NaN                  NaN   

                    oxygen_conc(mg/l)  
date                                   
1970-01-01 00:00:00               NaN  
1970-01-01 01:00:00               NaN  
1970-01-01 02:00:00               NaN  
1970-01-01 03:00:00               NaN  
1970-01-01 04:00:00               NaN  
...                               ...  
2022-12-30 20:00:00               NaN  
2022-12-30 21:00:00               NaN  
2022-12-30 22:00:00               NaN  
2022-12-30 23:00:00               NaN  
2022-12-31 00:00:00               NaN  

[464569 rows x 4 columns]

Station 2290

In [119]:
namestation = "2290"

columns_names = ["temperature(°C)", "pH(-)", "electric_cond(µS/cm)", "oxygen_conc(mg/l)"]
hourly_timeseries = pd.DataFrame(index = pd.date_range('01-01-1970','12-31-2022', freq='H'), columns=columns_names)

############ Electrical conductivity
path_2290_EC = 'data\S2290\S2290_St-Sulpice-Areuse_LF_20090425-20221231.CSV'

network_2290_EC = pd.read_csv(path_2290_EC, skiprows=13, delimiter = ";", encoding='latin-1')

network_2290_EC = network_2290_EC[["Datum.1", "Uhrzeit.1", "LF Stundenmittel [µS/cm]"]]

# Drop rows with NaN values in the specific column
network_2290_EC.dropna(subset=['Datum.1'], inplace=True)

network_2290_EC["date"] = network_2290_EC["Datum.1"] + " "  + network_2290_EC["Uhrzeit.1"]
network_2290_EC["date"] = pd.to_datetime(network_2290_EC["date"], format='%d.%m.%Y %H:%M:%S')
network_2290_EC.set_index("date", inplace = True)

############ Temperature
path_2290_T = 'data\S2290\S2290_St-Sulpice-Areuse_T_20090425-20221231.CSV'

network_2290_T = pd.read_csv(path_2290_T, skiprows=13, delimiter = ";", encoding='latin-1')

network_2290_T = network_2290_T[["Datum.1", "Uhrzeit.1", "T Stundenmittel [°C]"]]

# Drop rows with NaN values in the specific column
network_2290_T.dropna(subset=['Datum.1'], inplace=True)

network_2290_T["date"] = network_2290_T["Datum.1"] + " "  + network_2290_T["Uhrzeit.1"]
network_2290_T["date"] = pd.to_datetime(network_2290_T["date"], format='%d.%m.%Y %H:%M:%S')
network_2290_T.set_index("date", inplace = True)

############################################################################
network_2290_EC = network_2290_EC[["LF Stundenmittel [µS/cm]"]]
hourly_timeseries["electric_cond(µS/cm)"] = network_2290_EC["LF Stundenmittel [µS/cm]"]

network_2290_T = network_2290_T[["T Stundenmittel [°C]"]]
hourly_timeseries["temperature(°C)"] = network_2290_T["T Stundenmittel [°C]"]

hourly_timeseries.index.name = "date"
    
hourly_timeseries = hourly_timeseries.round(2)

hourly_timeseries.to_csv(PATH_OUTPUT + "/hourly/CAMELS_CH_chem_hourly_"+namestation+".csv", encoding='latin')

hourly_timeseries

temperature(°C) pH(-) electric_cond(µS/cm)  \
date                                                             
1970-01-01 00:00:00             NaN   NaN                  NaN   
1970-01-01 01:00:00             NaN   NaN                  NaN   
1970-01-01 02:00:00             NaN   NaN                  NaN   
1970-01-01 03:00:00             NaN   NaN                  NaN   
1970-01-01 04:00:00             NaN   NaN                  NaN   
...                             ...   ...                  ...   
2022-12-30 20:00:00            7.56   NaN                404.6   
2022-12-30 21:00:00            7.56   NaN                403.4   
2022-12-30 22:00:00            7.55   NaN                  403   
2022-12-30 23:00:00            7.55   NaN                  403   
2022-12-31 00:00:00            7.54   NaN                403.1   

                    oxygen_conc(mg/l)  
date                                   
1970-01-01 00:00:00               NaN  
1970-01-01 01:00:00               NaN  
1970-01-01 02:00:00               NaN  
1970-01-01 03:00:00               NaN  
1970-01-01 04:00:00               NaN  
...                               ...  
2022-12-30 20:00:00               NaN  
2022-12-30 21:00:00               NaN  
2022-12-30 22:00:00               NaN  
2022-12-30 23:00:00               NaN  
2022-12-31 00:00:00               NaN  

[464569 rows x 4 columns]

# End